File: add-two-numbers/solution2.go
// https://leetcode.com/problems/add-two-numbers/

package main

import (
	"fmt"
	"strconv"
	"strings"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

// T: O(max(m, n)), S: O(max(m, n))
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	fake := &ListNode{}
	tail := fake
	carry := 0
	for l1 != nil || l2 != nil {
		sum := carry
		if l1 != nil {
			sum += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			sum += l2.Val
			l2 = l2.Next
		}
		carry = sum / 10
		tail.Next = &ListNode{Val: sum % 10}
		tail = tail.Next
	}
	if carry > 0 {
		tail.Next = &ListNode{Val: carry}
	}
	return fake.Next
}

func (ln *ListNode) String() string {
	var sb strings.Builder
	for ln != nil {
		sb.WriteString(strconv.Itoa(ln.Val))
		if ln.Next != nil {
			sb.WriteString(" -> ")
		}
		ln = ln.Next
	}
	return sb.String()
}

func main() {
	l1 := &ListNode{Val: 2}
	l1.Next = &ListNode{Val: 4}
	l1.Next.Next = &ListNode{Val: 3}
	l2 := &ListNode{Val: 5}
	l2.Next = &ListNode{Val: 6}
	l2.Next.Next = &ListNode{Val: 4}
	l := addTwoNumbers(l1, l2)
	fmt.Println(l) // 7 -> 0 -> 8
}

File: add-two-numbers/solution2.ts
// https://leetcode.com/problems/add-two-numbers/

class ListNode {
    val: number;
    next: ListNode | null;
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val === undefined) ? 0 : val;
        this.next = (next === undefined) ? null : next;
    }
}

// T: O(max(m, n)), S: O(max(m, n))
function addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {
    const fake: ListNode = new ListNode(-1);
    let tail: ListNode = fake;
    let carry: number = 0;
    while (l1 || l2) {
        let sum: number = carry;
        if (l1) {
            sum += l1.val;
            l1 = l1.next;
        }
        if (l2) {
            sum += l2.val;
            l2 = l2.next;
        }
        carry = Math.floor(sum / 10);
        tail.next = new ListNode(sum % 10);
        tail = tail.next
    }
    if (carry) {
        tail.next = new ListNode(carry);
    }
    return fake.next;
}

function printList(node: ListNode | null): void {
    const values: number[] = [];
    while (node) {
        values.push(node.val);
        node = node.next;
    }
    console.log(values.join(' -> '));
}

let l1 = new ListNode(2, new ListNode(4, new ListNode(3)));
let l2 = new ListNode(5, new ListNode(6, new ListNode(4)));
printList(addTwoNumbers(l1, l2)); // 7 -> 0 -> 8

l1 = new ListNode(0);
l2 = new ListNode(0);
printList(addTwoNumbers(l1, l2)); // 0

l1 = new ListNode(9, new ListNode(9, new ListNode(9, new ListNode(9, new ListNode(9, new ListNode(9, new ListNode(9)))))));
l2 = new ListNode(9, new ListNode(9, new ListNode(9, new ListNode(9))));
printList(addTwoNumbers(l1, l2)); // 8 -> 9 -> 9 -> 9 -> 0 -> 0 -> 0 -> 1
